Add test_linux_watchdog: Test if the watchdog resets the device

SmithChart opened the pull request.

SmithChart committed:
Add test_linux_watchdog: Test if the watchdog resets the device

Signed-off-by: Chris Fiege <cfi@pengutronix.de>
Co-authored-by: Bastian Krause <bst@pengutronix.de>

SmithChart committed:
Add test_linux_watchdog: Test if the watchdog resets the device

Signed-off-by: Chris Fiege <cfi@pengutronix.de>
Co-authored-by: Bastian Krause <bst@pengutronix.de>

Bastian-Krause commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,35 @@ def test_sensors(shell, record_property):
+        # With any exception happening in this test we must assume that the device state is tainted.
+        # Let's switch it off, so the strategy can reboot the device into a clean state
+        strategy.transition(Status.off)
+        raise e
```
```suggestion
    finally:
        # Let's switch the TAC off, so the strategy can boot the device into a clean state
        strategy.transition("off")
```
..and drop the `from lxatacstrategy import Status`.

SmithChart commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,35 @@ def test_sensors(shell, record_property):
+            timeout=30,
+        )
+        if index != 0:
+            raise Exception("Device failed to reboot in time.")
```
Nothing. I guess this is just me misinterpreting the API. Applied.

SmithChart commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,35 @@ def test_sensors(shell, record_property):
+        strategy.barebox.run_check("global linux.bootargs.loglevel=loglevel=6")
+        strategy.barebox.boot("")
+        strategy.barebox.await_boot()
+        strategy.target.activate(strategy.shell)
```
That's right. Thanks for catching that!

SmithChart commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,35 @@ def test_sensors(shell, record_property):
     assert 10 <= data["cpu_thermal-virtual-0"]["temp1"]["temp1_input"] <= 70
+
+
+def test_linux_watchdog(strategy):
```
Thx, I've added the `shell` fixture.

SmithChart committed:
Environment: Use FQDN for remotelab server

All other machines referenced are already given with their FQDN.
So let's do the same for the remotelab server.

This change makes the environment portable to hosts where
`CanonicalizeHostname` is not present in the SSH configuration.

Signed-off-by: Chris Fiege <cfi@pengutronix.de>

SmithChart committed:
Environment: Use FQDN for remotelab server

All other machines referenced are already given with their FQDN.
So let's do the same for the remotelab server.

This change makes the environment portable to hosts where
`CanonicalizeHostname` is not present in the SSH configuration.

Signed-off-by: Chris Fiege <cfi@pengutronix.de>

SmithChart commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,35 @@ def test_sensors(shell, record_property):
+        strategy.target.deactivate(strategy.shell)
+
+        # systemd should be feeding the watchdog. let's kill systemd and wait for the watchdog to reset the DUT.
+        strategy.console.write(b"kill -11 1\n")  # This command will not return, so we can not use shell.run()
```
Interesting approach. This seems to be timing-sensitive. At least in my setup a `sleep 1` did not reliably print a prompt before more output is printed to the serial.

SmithChart commented:
@Bastian-Krause Sorry for letting this PR slide for so long and thanks for the feedback. With all your suggestions I really like how the test code turned out! Please have another look.

Bastian-Krause commented on tests/test_linux_hardware.py:
```diff
@@ -1,5 +1,9 @@
 
+import labgrid.driver
+
+from lxatacstrategy import LXATACStrategy
```
Hm, I'm not sure if using implicit namespace packages is a good idea..

SmithChart replied:
What is your suggestion to access the strategy here? I'll probably need the module containing the strategy for the other comment below anyway.

Bastian-Krause replied:
I can't think of anything better for test suites that aren't actual Python packages. Let's keep it this way.

Bastian-Krause replied:
The downside of this is that the test module is now tied to the specific strategy. If we would introduce a second one (for whatever reason), this does not work anymore..

SmithChart replied:
That's a good point. I would be fine to accept this for now and find a better solution if we ever get to this problem.

Bastian-Krause commented on tests/test_linux_hardware.py:
```diff
@@ -80,3 +84,26 @@ def test_sensors(shell, record_property):
+    finally:
+        # Inform the strategy that we are not in state "shell" any more.
+        # This way clean-up actions before power off do not run.
+        strategy.status = "unknown"
```
Don't we need an enum here?

SmithChart replied:
This probably works because in the strategy we only check for the new state when deciding how to transition to `off`. Using the Status-enum is probably cleaner.

Maybe you can elaborate why you voted against using the value from the enum [here](https://github.com/linux-automation/labgrid-lxatac/pull/18#discussion_r2514337061)?

Bastian-Krause replied:
The idea was to get rid of the import, but as discussed above, we need the import for the type hints. So we can use the import from the strategy here, too. Another option would be to make `transition("unknown") ` work.
